aptos-move/aptos-vm/src/zkid_validation.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::move_vm_ext::{AptosMoveResolver, SessionExt};
+use aptos_types::{
+    jwks::{jwk::JWK, PatchedJWKs},
+    on_chain_config::{CurrentTimeMicroseconds, OnChainConfig},
+    transaction::SignedTransaction,
+    vm_status::{StatusCode, VMStatus},
+    zkid::{ZkpOrOpenIdSig, MAX_ZK_ID_AUTHENTICATORS_ALLOWED},
+};
+use aptos_vm_logging::log_schema::AdapterLogSchema;
+use move_binary_format::errors::Location;
+use move_core_types::{language_storage::CORE_CODE_ADDRESS, move_resource::MoveStructType};
+
+fn get_current_time_onchain(
+    resolver: &impl AptosMoveResolver,
+) -> anyhow::Result<CurrentTimeMicroseconds, VMStatus> {
+    CurrentTimeMicroseconds::fetch_config(resolver).ok_or_else(|| {
+        VMStatus::error(
+            StatusCode::VALUE_DESERIALIZATION_ERROR,
+            Some("could not fetch CurrentTimeMicroseconds on-chain config".to_string()),
+        )
+    })
+}
+
+fn get_jwks_onchain(resolver: &impl AptosMoveResolver) -> anyhow::Result<PatchedJWKs, VMStatus> {
+    let error_status = VMStatus::error(
+        StatusCode::VALUE_DESERIALIZATION_ERROR,
+        Some("could not fetch PatchedJWKs".to_string()),
+    );
+    let bytes = resolver
+        .get_resource(&CORE_CODE_ADDRESS, &PatchedJWKs::struct_tag())
+        .map_err(|e| e.finish(Location::Undefined).into_vm_status())?
+        .ok_or_else(|| error_status.clone())?;
+    let jwks = bcs::from_bytes::<PatchedJWKs>(&bytes).map_err(|_| error_status.clone())?;
+    Ok(jwks)
+}
+
+pub fn validate_zkid_authenticators(
+    transaction: &SignedTransaction,
+    resolver: &impl AptosMoveResolver,
+    _session: &mut SessionExt,
+    _log_context: &AdapterLogSchema,
+) -> anyhow::Result<(), VMStatus> {
+    // TODO(ZkIdGroth16Zkp): The ZKP/OpenID sig verification does not charge gas. So, we could have DoS attacks.
+    let zkid_authenticators =
+        aptos_types::zkid::get_zkid_authenticators(transaction).map_err(|_| {
+            VMStatus::error(
+                StatusCode::INVALID_SIGNATURE,
+                Some("Failed to fetch zkid authenticators".to_owned()),
+            )
+        })?;
+
+    if zkid_authenticators.is_empty() {
+        return Ok(());
+    }
+
+    if zkid_authenticators.len() > MAX_ZK_ID_AUTHENTICATORS_ALLOWED {
```

Comment:
Why do we set a limit on the max number of authenticators?

Reply:
They are much more expensive to verify compared to a normal authenticator (Groth16 ZKP vs Ed25519; 2ms versus 40 microseconds)